rollout/trafficrouting/appmesh/appmesh.go

```diff
@@ -0,0 +1,386 @@
+	requiresUpdate := false
+
+	routesFilterMap := make(map[string]bool)
+	for _, r := range rRoutes {
+		routesFilterMap[r] = true
+	}
+
+	routesFldPath := field.NewPath("spec", "routes")
+	routesI, found, err := unstructured.NestedSlice(uVrCopy.Object, "spec", "routes")
+	if !found || err != nil {
+		return field.Invalid(routesFldPath, uVrCopy.GetName(), fmt.Sprintf("No routes found"))
+	}
+
+	for idx, routeI := range routesI {
+		routeFldPath := routesFldPath.Index(idx)
+		route, ok := routeI.(map[string]interface{})
+		if !ok {
+			return field.Invalid(routeFldPath, uVrCopy.GetName(), ErrNotWellFormed)
+		}
+
+		reconCtx := &routeReconcileContext{
+			route:           route,
+			routeIndex:      idx,
+			routeFldPath:    routeFldPath,
+			rCanaryVnodeRef: rCanaryVnodeRef,
+			rStableVnodeRef: rStableVnodeRef,
+			routesFilterMap: routesFilterMap,
+			desiredWeight:   desiredWeight,
+		}
+		ru, err := r.reconcileRoute(ctx, uVrCopy, reconCtx)
+		if err != nil {
+			return err
+		}
+		requiresUpdate = requiresUpdate || ru
+	}
+
+	//update virtual-router with updated routes
+	err = unstructured.SetNestedSlice(uVrCopy.Object, routesI, "spec", "routes")
+	if err != nil {
+		return err
+	}
+	if requiresUpdate {
+		_, err = r.client.UpdateVirtualRouterCR(ctx, uVrCopy)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileRoute(ctx context.Context, uVr *unstructured.Unstructured, routeCtx *routeReconcileContext) (bool, error) {
+	routeName, ok := routeCtx.route["name"].(string)
+	if !ok {
+		return false, field.Invalid(routeCtx.routeFldPath.Child("name"), uVr.GetName(), ErrNotWellFormed)
+	}
+
+	if len(routeCtx.routesFilterMap) > 0 {
+		// filter out the routes that are not specified in route filter
+		if _, ok := routeCtx.routesFilterMap[routeName]; !ok {
+			return false, nil
+		}
+	}
+
+	routeRule, routeType, err := GetRouteRule(routeCtx.route)
+	if err != nil && routeRule == nil {
+		return false, field.Invalid(routeCtx.routeFldPath, uVr.GetName(), ErrNotWellFormed)
+	}
+
+	weightedTargetsFldPath := routeCtx.routeFldPath.Child(routeType).Child("action").Child("weightedTargets")
+	weightedTargets, found, err := unstructured.NestedSlice(routeRule, "action", "weightedTargets")
+	if !found || err != nil {
+		return false, field.Invalid(weightedTargetsFldPath, uVr.GetName(), ErrNotWellFormed)
+	}
+
+	requiresUpdate := false
+	for idx, wtI := range weightedTargets {
+		wtFldPath := weightedTargetsFldPath.Index(idx)
+		wt, ok := wtI.(map[string]interface{})
+		if !ok {
+			return false, field.Invalid(wtFldPath, uVr.GetName(), ErrNotWellFormed)
+		}
+		wtVnRefFldPath := wtFldPath.Child("virtualNodeRef")
+		wtVnRef, ok := wt["virtualNodeRef"].(map[string]interface{})
+		if !ok {
+			return false, field.Invalid(wtVnRefFldPath, uVr.GetName(), ErrNotWellFormed)
+		}
+		wtVnName, _ := wtVnRef["name"].(string)
+		wtVnNamespace := defaultIfEmpty(wtVnRef["namespace"], r.rollout.Namespace)
+		weight, err := toInt64(wt["weight"])
```

Comment:
why do you need this be Int64?

Reply:
AppMesh CRD uses int64 in [here](https://aws.github.io/aws-app-mesh-controller-for-k8s/reference/api_spec/#appmesh.k8s.aws/v1beta2.WeightedTarget), hence needed to do this.